File: backend/src/application/use-cases/shopping-session/SaveShoppingSessionUseCase.ts
import { v4 as uuidv4 } from "uuid";
import type { SaveShoppingSessionUseCasePort } from "@application/ports/incoming/SaveShoppingSessionUseCasePort";
import type { ShoppingSessionRepository } from "@application/ports/outgoing/ShoppingSessionRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type {
	SaveShoppingSessionInput,
	SaveShoppingSessionResult,
} from "./contracts";
import { ShoppingSessionTotalCalculator } from "@domain/services/ShoppingSessionTotalCalculator";
import type { ShoppingSession } from "@domain/entities/ShoppingSession";

export class SaveShoppingSessionUseCase implements SaveShoppingSessionUseCasePort {
	constructor(
		private readonly shoppingSessionRepository: ShoppingSessionRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(input: SaveShoppingSessionInput): Promise<SaveShoppingSessionResult> {
		this.logger.info(
			`[SaveShoppingSessionUseCase] execute - userId: "${input.userId}", items: ${input.items.length}`,
		);

		if (input.items.length === 0) {
			throw new Error("Cannot save a shopping session with no items.");
		}

		const shoppedAt = new Date(input.shoppedAt);
		if (isNaN(shoppedAt.getTime())) {
			throw new Error(`Invalid shoppedAt date: "${input.shoppedAt}"`);
		}

		const totalPrice = ShoppingSessionTotalCalculator.calculate(input.items);
		const sessionId = uuidv4();

		const session: ShoppingSession = {
			id: sessionId,
			userId: input.userId,
			shoppedAt,
			totalPrice,
			createdAt: new Date(),
			items: input.items.map((item) => ({
				productName: item.productName,
				supermarket: item.supermarket,
				category: item.category,
				price: item.price,
				pricePerUnit: item.pricePerUnit,
				unit: item.unit,
				taxType: item.taxType,
				quantity: item.quantity,
				image: item.image,
				url: item.url,
			})),
		};

		const savedSession = await this.shoppingSessionRepository.save(session);

		this.logger.info(
			`[SaveShoppingSessionUseCase] session saved - id: "${savedSession.id}", total: ${savedSession.totalPrice}`,
		);

		return {
			sessionId: savedSession.id,
			totalPrice: savedSession.totalPrice,
			itemCount: savedSession.items.length,
		};
	}
}

File: backend-dotnet/src/api/Application/UseCases/ShoppingSession/SaveShoppingSessionUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Entities;
using SupermarketPlanner.Api.Domain.Services;

namespace SupermarketPlanner.Api.Application.UseCases.ShoppingSession;

public sealed class SaveShoppingSessionUseCase(
    IShoppingSessionRepository shoppingSessionRepository,
    ILogger<SaveShoppingSessionUseCase> logger) : ISaveShoppingSessionUseCase
{
    public Task<Either<DomainError, SaveShoppingSessionResult>> Invoke(
        SaveShoppingSessionInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[SaveShoppingSessionUseCase] Invoke - userId: \"{UserId}\", items: {Count}",
            input.UserId, input.Items.Count);

        if (input.Items.Count == 0)
            return Task.FromResult(Either<DomainError, SaveShoppingSessionResult>.FromLeft(
                DomainError.Validation(nameof(input.Items), "Cannot save a shopping session with no items.")));

        if (!DateTimeOffset.TryParse(input.ShoppedAt, out var shoppedAt))
            return Task.FromResult(Either<DomainError, SaveShoppingSessionResult>.FromLeft(
                DomainError.Validation(nameof(input.ShoppedAt), $"Invalid shoppedAt date: \"{input.ShoppedAt}\"")));

        var totalPrice = ShoppingSessionTotalCalculator.Calculate(
            input.Items.Select(item => (item.Price, item.Quantity)).ToList());

        var session = new ShoppingSessionEntity(
            Id: Guid.NewGuid().ToString(),
            UserId: input.UserId,
            ShoppedAt: shoppedAt,
            TotalPrice: totalPrice,
            CreatedAt: DateTimeOffset.UtcNow,
            Items: input.Items.Select(MapInputItemToSessionItem).ToList());

        return shoppingSessionRepository.Save(session, cancellationToken)
            .MapAsync(savedSession =>
            {
                logger.LogInformation(
                    "[SaveShoppingSessionUseCase] session saved - id: \"{SessionId}\", total: {Total}",
                    savedSession.Id, savedSession.TotalPrice);
                return new SaveShoppingSessionResult(
                    SessionId: savedSession.Id,
                    TotalPrice: savedSession.TotalPrice,
                    ItemCount: savedSession.Items.Count);
            });
    }

    private static ShoppingSessionItem MapInputItemToSessionItem(SaveShoppingSessionItemInput item) =>
        new(item.ProductName, item.Supermarket, item.Category, item.Price,
            item.PricePerUnit, item.Unit, item.TaxType, item.Quantity, item.Image, item.Url);
}
